### PR TITLE
Auto-scaling (resolves #349, resolves #600)

### LINKED ISSUE #349: Design and implement autoscaling algorithm

### LINKED ISSUE #600: Implement provisioner for EC2

### PR DESCRIPTION
**Pull request by benedictpaten**
Not finished, starting to add abstract API. Have sketched out an algorithm, will try to get this implemented today and tomorrow in a fairly dumb way.

Resolves #349

### CODE REVIEW

File: src/toil/provisioner/abstractProvisioner.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    """
+    def __init__( self, message ):
+        super( ProvisioningException, self ).__init__( message )
+
+class AbstractProvisioner(object):
+    """An abstract base class to represent the interface for provisioning 
+    worker nodes to use in a toil cluster.
+    """
+    
+    def addNodes(self, nodes=1, preemptable=False):
```
Comment (hannes-ucsc) [inline src/toil/provisioner/abstractProvisioner.py]: Use @abstractmethod

File: src/toil/provisioner/abstractProvisioner.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        """
+        Adds worker node to the set of worker nodes.
+        
+        :param int nodes: Number of nodes to added.
+        
+        :raise ProvisioningException: If worker nodes can not be added.
+        """
+        raise NotImplementedError('Abstract method: addNode')
```
Comment (hannes-ucsc) [inline src/toil/provisioner/abstractProvisioner.py]: No need for the constructor argument.

File: src/toil/provisioner/abstractProvisioner.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+# limitations under the License.
+
+class ProvisioningException( Exception ):
+    """
+    General provisioning exception. 
+    """
+    def __init__( self, message ):
+        super( ProvisioningException, self ).__init__( message )
+
+class AbstractProvisioner(object):
```
Comment (hannes-ucsc) [inline src/toil/provisioner/abstractProvisioner.py]: `__metaclass__ = ABCMeta`

File: src/toil/provisioner/abstractProvisioner.py (hunk 2 above)
Comment (hannes-ucsc) [inline src/toil/provisioner/abstractProvisioner.py]: ABCMeta catches unimplemented methods at construction time.

And even if it didn't, the backtrace would be informative enough.

**Comment by hannes-ucsc** [comment]
My hunch is that `getIssuedQueueSize()` sort of hard-wires one particular notion of autoscaling algorithm, the naive one, so to speak. I think we should just pass the provisioner into the batch system and have it make its own decisions. We can always generalize later and pull functionality into leader.py. It is easier to add abstraction than to change or remove them. Besides, the only batch system that could currently support auto-scaling is Mesos.

**Comment by hannes-ucsc** [comment]
Jenkins, skip this.

**Comment by hannes-ucsc** [comment]
Jenkins, skip this please.

**Comment by hannes-ucsc** [comment]
Build fails with straight-forward error. Multiple inheritance complicates the super.__init__ mechanism.

It is uncommon for an abstract base class (ABC) to introduce state. That should be left to the concrete subclasses. I think we should push the common attributes down into each concrete subclass.  If you do that the constructor of the ABC looses its arguments or maybe even becomes empty at which point you can drop it.

**Comment by benedictpaten** [comment]
Okay, that was silly, do you need me to take a look, or are you proceeding?

**Comment by hannes-ucsc** [comment]
I won't get to it before Monday. At that time I will be happy to fix it.

On Fri, Dec 4, 2015 at 3:14 PM, Benedict Paten <notifications@github.com>
wrote:

> Okay, that was silly, do you need me to take a look, or are you proceeding?
>
> On Fri, Dec 4, 2015 at 2:25 PM, Hannes Schmidt <notifications@github.com>
> wrote:
>
> > Build fails with straight-forward error. Multiple inheritance complicates
> > the super.*init* mechanism.
> >
> > It is uncommon for an abstract base class (ABC) to introduce state. That
> > should be left to the concrete subclasses. I think we should push the
> > common attributes down into each concrete subclass. If you do that the
> > constructor of the ABC looses its arguments or maybe even becomes empty
> at
> > which point you can drop it.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/BD2KGenomics/toil/pull/541#issuecomment-162099207>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BD2KGenomics/toil/pull/541#issuecomment-162107445>.
>



-- 
Hannes Schmidt
Software Engineer
Center for Biomolecular Science and Engineering
University of California, Santa Cruz

(206) 696-2316 (cell)
hannes@ucsc.edu

**Comment by hannes-ucsc** [comment]
Continued in #802

**Pull request closed without merging**